synapse/handlers/federation_event.py

```diff
@@ -1008,6 +1008,8 @@ async def _process_received_pdu(
             sender_key = event.content.get("sender_key")
 
             cached_devices = await self._store.get_cached_devices_for_user(event.sender)
+            if cached_devices is None:
+                cached_devices = {}
```

Comment:
(Based on the comment above in `devicemessage.py`...) is this expected?

Reply:
I don't think so, but I'm not all that sure to be honest. I wanted to at least err on the side of maintaining the existing behaviour. Similarly on line 558 below.

Reply:
This is gone too after the revert. `get_cached_devices_for_user` will return an empty dict
- if we've never requested this user's devices,
- or if we have, but their HS returned an empty device list.

Instead, we distinguish between these two cases at the point where we check to see if the cache needs updating.